### ISSUE TITLE
Chunked uploads create multiple files

### ISSUE THREAD

**Issue by cburatto** [balderdashy/skipper#51]
While using the skipper-disk adapter with a chunked upload, the server creates multiple files, instead of appending to the existing parts. Is there any option to deal with this?

**Comment by mikermcneil** [comment]
@cburatto Haven't been able to replicate-- would you mind providing a test case?  Also, can you try out skipper@0.5.4 and skipper-disk@0.5.4?

**Comment by cburatto** [comment]
Kind of in a hurry, let me give some basic info:
Server: sails 0.10.5
updated skipper to 0.5.4
skipper-disk 0.5.4
(please note since skipper is installed by default, I am not using  "require('skipper')", but updated skipper in package.json" -- I hope it s catching it)

Client:
Using Blueimp with chunked upload
https://github.com/blueimp/jQuery-File-Upload/wiki/Chunked-file-uploads
maxChunkSize: 10000000

Upload a file that is larger than maxChunkSize (in my case, 14 MB)

Result:
Check .tmp/uploads
Server has created 2 uuid files: one with 10000000 bytes, the other with 4382088 bytes

I didn't have the time to check with other chunked clients -- not sure if it could be a Blueimp issue.

Sorry for the hurry. I'll try to run other tests.

Thanks for the great work with Sails. I am in love with it.

**Comment by mikermcneil** [comment]
@cburatto thanks, that gives me a good place to start from.  Haven't seen this before, but I also haven't used that particular setting; so I'll try and replicate.  If you have a sec, can you drop your front-end code in a repo?  No worries about automated tests yet.  I wonder if it's triggering the controller action twice?

**Comment by cburatto** [comment]
Sorry for the delay: busy as hell. I've been checking this in more details and the thing is that, although the browser is able to  "chunk" the files, the server has to be aware and reassemble the chunks. So the Skipper adapters should be prepared to check this possibility and act accordingly.

It seems the most common technique used is by reading the HTTP headers:

``Content-Type``
``Content-Range``
``Content-Disposition``

This is what blueimp/jQuery-File-Upload does (you can check his PHP server here: https://github.com/blueimp/jQuery-File-Upload/blob/master/server/php/UploadHandler.php ) and what https://github.com/mailru/FileAPI does as well:
>
Client and server communicate to each other using the following HTTP headers and status codes.
Client explicitly sets the following headers:
>
Content-Range: bytes <start-offset>-<end-offset>/<total>
Content-Disposition: attachment; filename=<file-name>
Any other headers are set by a target browser and are not used by client. Library does not provide any facilities to track a file uniqueness across requests, it's left on developer's consideration.
Response codes:
>
200 - last chunk is uploaded
201 - chunk is successfully saved
416 - range is not acceptable error, recoverable
500 - server error, recoverable

I am not sure if this is a standard. For example, if you check what Resumable does:

http://www.resumablejs.com/

You will see that it relies on extra client parameters that the server should use for appending: ``resumableChunkNumber``, ``resumableChunkSize``, etc. 

Since there is no way to create a default Skipper adapter for every client, I would suggest that Skipper used ``Content-Type``, ``Content-Range``, ``Content-Disposition`` to check if the incoming file is chunked and decide if this is a chunked upload or not. Any other custom implementation would require some custom adapter. But for the sake of loose coupling and making sure the client can be anything, not only a browser, defaulting to the headers should be a best practice.

I would try implementing this, but I am really a newbie and would just take forever and probably break everything in the way. 

I don't know how Skipper would keep track of the server file name (uuid or custom) to be appended, since chunked uploads can occur with a long time between retries (even between different machines, theoretically). 

And there is also the cluster scenario, when different servers could be picking up chunks.

Despite the complexity, I think being ready for chunked uploads would make it really robust. Maybe in chunked situations, Skipper should switch to filesystem storage, and only do S3 and GridFS after completion.

I'll post some front-end code next.

**Comment by sailsbot** [comment]
Thanks for posting, @cburatto.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
